### PR TITLE
replace shared-informers' methods with cached client

### PR DESCRIPTION
**Pull request by MatousJobanek**
[KUBESAW-132](https://issues.redhat.com/browse/KUBESAW-132)

First PR to replace the custom implementation of shared informers in reg-service with the cached controller-runtime client.
This PR changes the `informer_service.go`'s methods so they use the client instead.
I still keep the service file with the same methods to minimize the scope of the changes
* I don't need to change the places where it is used
* I can also update the unit tests in a separate PR

### CODE REVIEW

File: pkg/signup/service/signup_service_test.go [hunk 1]
```diff
@@ -273,8 +273,8 @@ func (s *TestSignupServiceSuite) TestGetSignupFailsWithNotFoundThenOtherError()
 		s.Application.MockInformerService(inf)
 		svc := service.NewSignupService(
 			fake.MemberClusterServiceContext{
-				Client: s,
-				Svcs:   s.Application,
+				CrtClient: s,
```
Comment (xcoulon) [inline pkg/signup/service/signup_service_test.go:276]: why do we need to rename the field to `CrtClient`?

**Review by xcoulon — APPROVED** [review]

File: pkg/signup/service/signup_service_test.go (hunk 1 above)
Comment (MatousJobanek) [reply pkg/signup/service/signup_service_test.go:276]: Because there is a new method called `Client` and you cannot have an exported field with the same name. At the end of the day (and this whole refactoring madness) this will all go away

Comment (xcoulon) [reply pkg/signup/service/signup_service_test.go:276]: ah right, I missed that

**Review by mfrancisc — APPROVED** [review]
Looks good ! 👍 

Thanks for the simplification 🙏 

I have only one query - do we have any data/idea about performance impact of this switch ? I mean will it be faster/same/slower or we're unsure ?

**Comment by MatousJobanek** [comment]
Great question @mfrancisc. 
We don't have any real data or document for this. But I've tested it locally, I created a few thousands of UserSignups, MURs, Spaces, and SpaceBindings and I tested the proxy - there was no delay.
But I could have tried to compare the numbers. Let me try that.

**Comment by mfrancisc** [comment]
I was mostly curious - maybe there are some docs/blog posts providing comparison/pros and cons between the two approaches.

**Comment by MatousJobanek** [comment]
As I see it, the biggest benefits/pros are:
* simplifies our code
* we don't need to maintain our own implementation of the shared informers
* the same is used in all our controllers
* we saw that the memory consumption is lower compared to our own custom implemenation

**Comment by MatousJobanek** [comment]
so I have some numbers, I created 16k of UserSignups, MURs, Spaces, and SpaceBindings (I meak 16k for each resource kind)
I made a request to get ConfigMaps from user's namespace via the proxy:
* with the old custom implementation it took around 60-90 milliseconds
* with the new one the number was the same
* without any cache it took 2.3-2.5 seconds

as for memory consumption of the reg-service pods:
* old custom implementation 700-800MiB
* new one 400-450MiB
* without cache around 35MiB

**Review by alexeykazakov — COMMENTED** [review]
We used controller runtime client(s) in our first proxy version as far as I remember. Assuming that the built-in client cache would be enough. But the performance was not consistent. Every few minutes of constant proxy usage the performance degraded. It looked like the cache was "reset" every few minutes or so.
This was the main reason why we switched to informers in the first place.
But maybe we did something different back then? If so then what exactly we did differently and why it should work now?
I might be missing or forgetting something though.
We also might need to do additional performance testing (or collaborate with the konflux folks to closely monitor any potential performance issues after deploying it and be ready to revert it if needed).

cc: @rajivnathan

**Comment by MatousJobanek** [comment]
I can test that - I still have the CRC with all the resources.

**Comment by alexeykazakov** [comment]
No, I'm not sure! Let me try to find the original code so we can compare.

**Comment by MatousJobanek** [comment]
I don't see any cache reset or performance difference after 15, 40, or 60 minutes:
```bash
$ oc get cm --token=XXX --server=https://api-toolchain-host-operator.apps-crc.testing --kubeconfig /tmp/dummy-kubeconfig --cache-dir /tmp/dummy-dir/ -v 6
I0918 08:33:54.361713 2595794 loader.go:373] Config loaded from file:  /tmp/dummy-kubeconfig
I0918 08:33:54.434089 2595794 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/api?timeout=32s 200 OK in 71 milliseconds
I0918 08:33:54.520956 2595794 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/apis?timeout=32s 200 OK in 79 milliseconds
I0918 08:33:54.656424 2595794 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/api/v1/namespaces/mjobanek-dev/configmaps?limit=500 200 OK in 97 milliseconds
NAME                       DATA   AGE
kube-root-ca.crt           1      15h
openshift-service-ca.crt   1      15h

# waited for 40 minutes

$ rm -rf /tmp/dummy-dir/
$ oc get cm --token=XXX --server=https://api-toolchain-host-operator.apps-crc.testing --kubeconfig /tmp/dummy-kubeconfig --cache-dir /tmp/dummy-dir/ -v 6
I0918 09:12:53.866156 2597468 loader.go:373] Config loaded from file:  /tmp/dummy-kubeconfig
I0918 09:12:53.908376 2597468 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/api?timeout=32s 200 OK in 41 milliseconds
I0918 09:12:53.929227 2597468 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/apis?timeout=32s 200 OK in 19 milliseconds
I0918 09:12:53.968135 2597468 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/api/v1/namespaces/mjobanek-dev/configmaps?limit=500 200 OK in 23 milliseconds
NAME                       DATA   AGE
kube-root-ca.crt           1      16h
openshift-service-ca.crt   1      16h

# waited for 15 minutes

$ rm -rf /tmp/dummy-dir/
$ oc get cm --token=XXX --server=https://api-toolchain-host-operator.apps-crc.testing --kubeconfig /tmp/dummy-kubeconfig --cache-dir /tmp/dummy-dir/ -v 6
I0918 09:18:20.803630 2598547 loader.go:373] Config loaded from file:  /tmp/dummy-kubeconfig
I0918 09:18:20.833579 2598547 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/api?timeout=32s 200 OK in 29 milliseconds
I0918 09:18:20.853703 2598547 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/apis?timeout=32s 200 OK in 19 milliseconds
I0918 09:18:20.895189 2598547 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/api/v1/namespaces/mjobanek-dev/configmaps?limit=500 200 OK in 26 milliseconds
NAME                       DATA   AGE
kube-root-ca.crt           1      16h
openshift-service-ca.crt   1      16h

# waited for more than 1 hour

$ rm -rf /tmp/dummy-dir/
$ oc get cm --token=XXX --server=https://api-toolchain-host-operator.apps-crc.testing --kubeconfig /tmp/dummy-kubeconfig --cache-dir /tmp/dummy-dir/ -v 6
I0918 10:21:25.334231 2604585 loader.go:373] Config loaded from file:  /tmp/dummy-kubeconfig
I0918 10:21:25.382830 2604585 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/api?timeout=32s 200 OK in 47 milliseconds
I0918 10:21:25.407047 2604585 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/apis?timeout=32s 200 OK in 23 milliseconds
I0918 10:21:25.452966 2604585 round_trippers.go:553] GET https://api-toolchain-host-operator.apps-crc.testing/api/v1/namespaces/mjobanek-dev/configmaps?limit=500 200 OK in 30 milliseconds
NAME                       DATA   AGE
kube-root-ca.crt           1      17h
openshift-service-ca.crt   1      17h

```

**Comment by MatousJobanek** [comment]
I realized one thing now, since I deleted the cache, then the `oc` did discovery calls, so the real time wasn't really visible for the "get configmaps" call. Let me do that one more time without resetting the cache.

**Comment by MatousJobanek** [comment]
There is actually the metric that should show the process time of the requests in proxy, but I have no clue how to visualize that easily in dev cluster.
Maybe @xcoulon knows. I guess that I would need to configure the Prometheus and ServiceMonitor properly, but I'm not sure if I want to go through the process - if it's worth it.

**Review by alexeykazakov — APPROVED** [review]
OK. In the original code we didn't use the controller-runtime cluster's client. We used a regular client instead. So I guess this is why you couldn't reproduce this problem with "cache reset" every few minutes.
Thank you for the extensive performance testing! I think it's good enough. And if we get complaints about degradation proxy performance then we can re-visit it. But for now I think we are good! :+1: 
I have just a couple of minor comments.
Great job!

File: cmd/main.go [hunk 2]
```diff
@@ -202,9 +191,22 @@ func newCachedClient(ctx context.Context, cfg *rest.Config) (client.Client, erro
...
+		"MasterUserRecord": &toolchainv1alpha1.MasterUserRecordList{},
+		"Space":            &toolchainv1alpha1.SpaceList{},
+		"SpaceBinding":     &toolchainv1alpha1.SpaceBindingList{},
+		"ToolchainStatus":  &toolchainv1alpha1.ToolchainStatusList{},
+		"UserSignup":       &toolchainv1alpha1.UserSignupList{},
+		"ProxyPlugin":      &toolchainv1alpha1.ProxyPluginList{},
+		"NSTemplateTier":   &toolchainv1alpha1.NSTemplateTierList{},
+		"ToolchainConfig":  &toolchainv1alpha1.ToolchainConfigList{},
+		"BannedUser":       &toolchainv1alpha1.BannedUserList{},
+		"Secret":           &corev1.SecretList{}}
```
[the remark this answers is not in the document]
Comment (rajivnathan) [reply cmd/main.go:204]: I believe Matous added these resources based on the informers we were creating in informers.go. If my memory serves me correctly, the list of resources there are those that are needed in the proxy flow and so we pre-populate the cache for those resources so that the proxy is as fast as possible even on first time use by a user. I don't remember whether ConfigMaps are part of that flow but I don't think it hurts to add it.

**Review by rajivnathan — APPROVED** [review]
Great work @MatousJobanek! 🙌  It's very cool to see the memory footprint improvement as well! 🤩 

>We used controller runtime client(s) in our first proxy version as far as I remember. Assuming that the built-in client cache would be enough. But the performance was not consistent. Every few minutes of constant proxy usage the performance degraded. It looked like the cache was "reset" every few minutes or so.
This was the main reason why we switched to informers in the first place.
But maybe we did something different back then? If so then what exactly we did differently and why it should work now?
I might be missing or forgetting something though.

@alexeykazakov TBH, I don't remember the reason why we used informers. Looking back in the [JIRA](https://issues.redhat.com/browse/CRT-1619) and the [code](https://github.com/codeready-toolchain/registration-service/blob/24e0696e1d8bed6a09d669a6ed84ee926d448317/pkg/proxy/cache.go#L51-L61) it sounds like we were using some simple cache that mapped users to cluster access (this part I recall) but that breaks when a user is retargeted to another cluster and that was an issue we were trying to solve. Not sure whether there were other issues that led us to use informers over runtime client. Maybe we just didn't realize that the runtime client is backed by informers at the time?

File: pkg/kubeclient/signup.go [hunk 3]
```diff
@@ -98,26 +99,29 @@ func (c *userSignupClient) listActiveSignupsByLabelForHashedValue(labelKey, valu
 // label matching the specified label
 func (c *userSignupClient) listActiveSignupsByLabel(labelKey, labelValue string) ([]*crtapi.UserSignup, error) {
 
-	stateRequirement, err := labels.NewRequirement(crtapi.UserSignupStateLabelKey, selection.NotEquals, []string{crtapi.UserSignupStateLabelValueDeactivated})
+	// TODO add unit tests checking that the label selection works as expected. Right now, it's not possible to do that thanks to the great abstraction and multiple level of layers, mocks, and services.
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/kubeclient/signup.go:102]: Yes, this is one big :man_facepalming:  
I was looking at options of covering that in unit-tests of the verification function, but unfortunately, it's not easily doable with the current setup.

File: cmd/main.go [hunk 4]
```diff
@@ -202,9 +191,22 @@ func newCachedClient(ctx context.Context, cfg *rest.Config) (client.Client, erro
 
 	// populate the cache backed by shared informers that are initialized lazily on the first call
 	// for the given GVK with all resources we are interested in from the host-operator namespace
-	objectsToList := []client.ObjectList{&toolchainv1alpha1.ToolchainConfigList{}, &corev1.SecretList{}}
-	for i := range objectsToList {
-		if err := hostCluster.GetClient().List(ctx, objectsToList[i], client.InNamespace(configuration.Namespace())); err != nil {
+	objectsToList := map[string]client.ObjectList{
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply cmd/main.go:194]: We list only those objects we want to cache upfront. As it is mentioned in the comment above this code, the cache works in a lazy way so it starts caching GVK resources only after it is fetched for the first time. 
In other words, the cache is populated only after making the first call on the GVK resource.

If we didn't do the list here, then the very first call to the reg-service/proxy would take a lot of time because the client would start populating the cache with the resources. 
By listing the resources before the reg-service pod gets ready, we ensure that the cache is fully synced with all resources we need to have cached.

File: cmd/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply cmd/main.go:204]: @rajivnathan is correct, we want to cache only those resources we access in reg-service. Caching anything else doesn't make much sense. 

We could theoretically drop a list of some reources like Secrets, NSTemplateTiers, ProxyPlugins, ToolchainStatuses, and ToolchainConfigs because we don't expect many resources to be present in the namespace, so the very first call wouldn't take much more time compared to when it is already cached. But I thought that it's also completely fine to pre-populate the cache with all resources the reg-service and proxy touch.

File: cmd/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply cmd/main.go:194]: I got that. My point is that it's easy to forget to add a new resource to this list. Let's say we are introducing some change to the proxy so it now needs to access a new resource with kind `X`. If we don't add it to this list then the proxy performance may degrade for the first request (it will degrade again after the proxy is restarted for every replica).
You can't access a new kind with adding this `X` resource to the operator role in https://github.com/codeready-toolchain/host-operator/blob/master/deploy/registration-service/registration-service.yaml#L20
So I was wondering if that yaml is a good place to add a reminder of updating the cache initialization if desired.

File: cmd/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply cmd/main.go:204]: Cache can be crucial for performance. Especially for the proxy. And it's easy to miss this (don't add specific resources to the initialization step) during development. We don't have any performance tests in our CI.
If neither reg-service or proxy touches CMs then IMO we should remove it from the role.
And I would argue that it would be safer to keep everything reg-service & proxy access in the cache initialization.

I checked it looks like the CM are not used by reg-service and proxy. So let's create a separate PR to remove it from the role.

File: cmd/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply cmd/main.go:194]: As discussed in the call:
1. I'll add a comment to the Role template 
2. I'll drop the ConfigMap from the Role
3. I'll try to make the list for loop a bit more generic so it goes over all toolchain kinds except for some specific ones like TierTemplate which we don't want to cache (there can be many of them and they are huge resources)

[the remark this answers is not in the document]
Comment (MatousJobanek) [reply cmd/main.go:194]: as for 1. and 2. see the PRs:
https://github.com/codeready-toolchain/host-operator/pull/1089
https://github.com/codeready-toolchain/registration-service/pull/463

As for the 3. item, it's easily doable to go over all kinds of the toolchain api, but there is one problem. The scheme contains api of both host & member, and at the level of the scheme we cannot decide which one is the host resource and which one belongs to the member. We could list available api groups from the cluster and exclude those kinds that are not available, but this wouldn't work in single-cluster environments.  In addition to that, host-operator SA doesn't have permission to read/list member-operator CRDs, so it would fail on authorization as well.
In other words, there are too many complications in doing it in a generic way so it's not worth it.
